clang/runtime/dpct-rt/include/dpct/device.hpp

```diff
@@ -732,6 +740,31 @@ class dev_mgr {
     select_device(selected_device_id);
   }
 
+  /// Update the device stack for the current thread id
+  void push_device(unsigned int id) {
+    auto tid = get_tid();
+    auto it = _thread2dev_map.find(tid);
+    if (it == _thread2dev_map.end())
+      _thread2dev_map[tid] = std::stack<unsigned int>();
+    _thread2dev_map[tid].push(id);
+  }
+
+  /// Remove the device from top of the stack if it exist
+  unsigned int pop_device() {
+    auto tid = get_tid();
+    auto it = _thread2dev_map.find(tid);
+    if (it == _thread2dev_map.end())
+      return -1;
+
+    if (it->second.empty())
+      return -1;
```

Comment:
Does CUDA always have a default context?

Reply:
I updated your code like below:
```
#include <cuda.h>
#include <iostream>


#define checkCUDAStatus(CALL) do {\
    cudaError_t Res = CALL;\
    if (Res != cudaSuccess) {\
        printf("Failed at %s:%d, error: %d\n", __FILE__, __LINE__, (int)Res);\
    }\
} while(0)

#define checkCUStatus(CALL) do {\
    CUresult Res = CALL;\
    if (Res != CUDA_SUCCESS) {\
        printf("Failed at %s:%d, error: %d\n", __FILE__, __LINE__, (int)Res);\
    }\
} while(0)

int main() {
  CUdevice  dev;
  CUcontext ctx;

  checkCUStatus(cuInit(0));

  checkCUDAStatus(cudaGetDevice(&dev));

  checkCUStatus(cuCtxGetCurrent(&ctx));
  std::cout << ctx << "\n";

  checkCUStatus(cuCtxCreate(&ctx, 0, dev));
  std::cout << "After create " << ctx << "\n";

  checkCUStatus(cuCtxPushCurrent(ctx));
  std::cout << "After push " << ctx << "\n";

  checkCUStatus(cuCtxPopCurrent(&ctx));
  std::cout << ctx << "\n";

  checkCUStatus(cuCtxPopCurrent(&ctx));
  std::cout << ctx << "\n";

  ctx = 0;

  checkCUStatus(cuCtxPopCurrent(&ctx));
  std::cout << ctx << "\n";

  checkCUStatus(cuCtxPopCurrent(&ctx));
  std::cout << ctx << "\n";

  checkCUStatus(cuDevicePrimaryCtxRetain(&ctx, dev));

  checkCUStatus(cuCtxDestroy(ctx));

  return 0;
}
```
```
zhiweij1@shliclel672:/export/users/zhiweij1/workspace$ nvcc a.cu -lcuda
zhiweij1@shliclel672:/export/users/zhiweij1/workspace$ ./a.out 
0
After create 0x55fd1eda3b30
After push 0x55fd1eda3b30
0x55fd1eda3b30
0x55fd1eda3b30
Failed at a.cu:44, error: 201
0
Failed at a.cu:47, error: 201
0
Failed at a.cu:52, error: 201
zhiweij1@shliclel672:/export/users/zhiweij1/workspace$ 
```
It seems that the cuda API will return a non-zero error code if we try to pop from an empty stack.

Reply:
So I agree you can return -1 here.